lib/src/voip/voip.dart

```diff
@@ -6,199 +6,293 @@ import 'package:webrtc_interface/webrtc_interface.dart';
-          Logs().v('[VOIP] onRoomState: type ${event.toJson()}.');
-          await onRoomStateChanged(event);
+          Logs().v('[VOIP] onRoomState: type ${event.toJson()}');
+          for (final map in groupCalls.entries) {
+            if (map.key.roomId == event.room.id) {
+              // because we don't know which call got updated, just update all
+              // group calls for that room
+              await map.value.onMemberStateChanged();
+            }
+          }
         }
       },
     );
 
-    client.onToDeviceEvent.stream.listen((event) async {
-      Logs().v('[VOIP] onToDeviceEvent: type ${event.toJson()}.');
+    delegate.mediaDevices.ondevicechange = _onDeviceChange;
+  }
 
-      if (event.type == 'org.matrix.call_duplicate_session') {
-        Logs().v('[VOIP] onToDeviceEvent: duplicate session.');
-        return;
-      }
+  Future<void> _handleCallEvents(List<BasicEventWithSender> callEvents) async {
+    // Call invites should be omitted for a call that is already answered,
+    // has ended, is rejectd or replaced.
+    final callEventsCopy = List<BasicEventWithSender>.from(callEvents);
+    for (final callEvent in callEventsCopy) {
+      final callId = callEvent.content.tryGet<String>('call_id');
+
+      if (CallConstants.callEndedEventTypes.contains(callEvent.type)) {
+        callEvents.removeWhere((event) {
+          if (CallConstants.ommitWhenCallEndedTypes.contains(event.type) &&
+              event.content.tryGet<String>('call_id') == callId) {
+            Logs().v(
+                'Ommit "${event.type}" event for an already terminated call');
+            return true;
+          }
+
+          return false;
+        });
+      }
+
+      // checks for ended events and removes invites for that call id.
+      if (callEvent is Event) {
+        // removes expired invites
+        final age = callEvent.unsigned?.tryGet<int>('age') ??
+            (DateTime.now().millisecondsSinceEpoch -
+                callEvent.originServerTs.millisecondsSinceEpoch);
+
+        callEvents.removeWhere((element) {
+          if (callEvent.type == EventTypes.CallInvite &&
+              age >
+                  (callEvent.content.tryGet<int>('lifetime') ??
+                      CallTimeouts.callInviteLifetime.inMilliseconds)) {
+            Logs().e(
+                '[VOIP] Ommiting invite event ${callEvent.eventId} as age was older than lifetime');
+            return true;
+          }
+          return false;
+        });
+      }
+    }
+
+    // and finally call the respective methods on the clean callEvents list
+    for (final callEvent in callEvents) {
+      await _callStreamByCallEvent(callEvent);
+    }
+  }
 
-      final confId = event.content['conf_id'];
-      final groupCall = groupCalls[confId];
-      if (groupCall == null) {
-        Logs().d('[VOIP] onToDeviceEvent: groupCall is null.');
+  Future<void> _callStreamByCallEvent(BasicEventWithSender event) async {
```

Comment:
I have no idea, what this function does based on its name. I would expect it to return a call stream, but it clearly does not :D